### PR TITLE
Create abstract Job interface block

### PR DESCRIPTION
**Pull request by ahuang11**
Starts implementing the abstract job block for Collections to use.

Represents a resource that can be executed in an external system. Often represents delegated orchestration by Prefect to another external system (Airbyte, Databricks, dbt Cloud, etc). Triggering execution, waiting for completion, and results retrieval are separated to allow for different methods of orchestration. That orchestration would be performed by a flow that accepts a block that implements this interface.

For example, a concrete, working example of JobBlock can be found in [`prefect_dbt.DbtCloudJob`](https://github.com/PrefectHQ/prefect-dbt/pull/101/files#diff-826f2d10c2bbb1fb21d3f9003e437a53594ab7b22f01e4a75b8a14ab7c9b4fedR540). and [DatabricksJob](https://github.com/PrefectHQ/prefect-databricks/pull/48)

### Example

image

### Checklist

- [ ] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [ ] This pull request includes tests or only affects documentation.
- [ ] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`

### CODE REVIEW

File: src/prefect/blocks/abstract.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+class JobRun(ABC):  # not a block
+    """
+    Represents a job run in an external system that can wait
+    for the job run's completion and fetch its results
+    """
+
+    def __init__(
+        self,
+        credentials: Block,
```
Comment (ahuang11) [inline src/prefect/blocks/abstract.py:18]: Not sure if this is ideal.

**Comment by ahuang11** [comment]
For an implementation of DatabaseBlock, see [`prefect_sqlalchemy.Database`](https://github.com/PrefectHQ/prefect-sqlalchemy/pull/35/files#diff-50da0519b78eac42cf22f4a901fd6582d3aba3f6f32f7789e1da3198924918a5R160)

File: src/prefect/blocks/abstract.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        If a run context is present, the logger property returns a run logger.
+        Else, it returns a default logger labeled with the class's name.
+        """
+        try:
+            return get_run_logger()
+        except MissingContextError:
+            return get_logger(self.__class__.__name__)
+
+    @abstractmethod
+    async def trigger(self, *args, **kwargs) -> JobRun:
```
Comment (desertaxle) [inline src/prefect/blocks/abstract.py:65]: Similar comment to the methods on `JobRun`
```suggestion
    async def trigger(self) -> JobRun:
```

**Review by zzstoatzz — APPROVED** [review]
nice!

File: src/prefect/blocks/abstract.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            return get_logger(self.__class__.__name__)
+
+    @abstractmethod
+    async def wait_for_completion(self):
+        """
+        Wait for the job run to complete.
+        """
+
+    @abstractmethod
+    async def fetch_results(self) -> T:
```
Comment (madkinsz) [inline src/prefect/blocks/abstract.py:38]: Why not `result()`?

Comment (ahuang11) [reply src/prefect/blocks/abstract.py:38]: Alex prefers verbs for method names; Nate suggests it muddies the concept of the Prefect future / result. Perhaps a compromise of `fetch_result`?

Comment (madkinsz) [reply src/prefect/blocks/abstract.py]: Not a hard feeling! I was just curious — I think that reasoning is solid. Having consistent names (i.e. verbs) sounds good to me.

File: src/prefect/blocks/abstract.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+from prefect.exceptions import MissingContextError
+from prefect.logging.loggers import get_logger
+
+T = TypeVar("T")
+
+
+class JobRun(ABC, Generic[T]):  # not a block
+    """
+    Represents a job run in an external system that can wait
+    for the job run's completion and fetch its results.
```
Comment (desertaxle) [inline src/prefect/blocks/abstract.py:15]:
```suggestion
    Represents a job run in an external system. Allows waiting
    for the job run's completion and fetching its results.
```

File: src/prefect/blocks/abstract.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    async def fetch_result(self) -> T:
+        """
+        Retrieve the results of the job run and return them.
+        """
+
+
+class JobBlock(Block, ABC):
+    """
+    Block that represents a resource that can trigger a long
+    running job in an external service.
```
Comment (desertaxle) [inline src/prefect/blocks/abstract.py:47]:
```suggestion
   Block that represents an entity in an external service that can trigger a long running execution.
```

**Review by desertaxle — APPROVED** [review]
I have two small nits on docstrings, but the code LGTM!

File: src/prefect/blocks/abstract.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    @abstractmethod
+    async def fetch_result(self) -> T:
+        """
+        Retrieve the results of the job run and return them.
+        """
+
+
+class JobBlock(Block, ABC):
+    """
+   Block that represents an entity in an external service that can trigger a long running execution.
```
Comment (ahuang11) [inline src/prefect/blocks/abstract.py:46]:
```suggestion
    Block that represents an entity in an external service that can trigger a long running execution.
```

**Pull request merged**